---
title: "HUGEN 2072 Project 1"
author: "Tianze (Vincent) Luo adapted"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: show
    toc: true
    toc_float: 
      collapsed: no
    df_print: paged
    number_sections: no
    theme: cosmo
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

```

# Instructions
Below is a tutorial consisting of a mostly-completed QC pipeline and some prompts.

Project 1 has three components:

* **Scripting/coding**: read through this file and complete the pipeline by contributing your code *where indicated*.

* **Narrative**: knit the completed .Rmd to a code-integrated HTML report of your QC process and decisions.

  * Your report must answer all the questions and perform all the tasks indicated in the instructions/prompts given for each step.

  * You must submit a successfully-knitted HTML file and the `.Rmd` in order to earn a passing grade for this project.

  * Set up your report to flow like an actual document you might show your supervisor to summarize the QC project, emphasizing what was done at each step rather than the code used to do it (e.g., how many samples/SNPs were filtered out at a given step?)

  * **Submit your .Rmd and HTML to Canvas**

* **Presentation**: you will make a 10 min (maximum) recording of yourself with Panopto in which you summarize this QC pipeline. Instructions will given on Canvas.

In each Part of this project you will implement one of the main steps of cleaning ***genotyping array data*** in a small dataset (note that we left a few steps out). You'll be shown how to use functions from the `{GWASTools}` package, and you'll supply a little of your own code to complete the pipeline that's been provided. At each step, you may be asked to create a filter for the data, to address errors/discrepancies, etc. Then for each step you'll make a short report (a few sentences and figures) explaining what you did.

Note that "sample" and "scan" are used interchangeably throughout this assignment, but that these are NOT synonymous with subject/participant/individual/etc. In this assignment, "SNP" and "probe" are used interchangeably, but in other contexts you should always remember the distinction.

**Since the pipeline below is incomplete, some of the R code chunks are set to `eval=FALSE` so that knitting to HTML will succeed. You will need to change them to `eval=TRUE` as you progress so that your knitted document shows the results of running each chunk.**

**It is expected that you will copy this file somewhere to your directory on the CRC cluster so that you can use the RStudio Server, where all the packages you need have been installed already.**


## Part 0 - Setup and Introduction to Annotation Data Frames

First, load the packages we'll be using.

```{r load_packages, message=F, eval=TRUE}

# Load the required packages for the project (Install them if necessary)
library("GWASTools", quietly = TRUE)
library("GWASdata", quietly = TRUE)
library("SNPRelate", quietly = TRUE)

library(tidyverse, quietly=TRUE)

# Close any open GDS files before proceeding (think of this as starting off with
# a 'blank slate')
showfile.gds(closeall = T, verbose = T)

```

The array data (genotypes, intensities, and BAF/LRR information) for this project are contained in several GDS files; accompanying SNP and scan annotation are also provided as an `.RData` file.

The probe/SNP annotation is stored in an object called a SNP annotation data frame, and the sample-level annotation is stored in a scan annotation data frame. Each of these objects is similar to an ordinary data frame. More precisely, a data frame is embedded within each. There is also accompanying metadata. 

To access the data and metadata, the `{GWASTools}` package provides several functions. Three basic functions you should familiarize yourself with are: 

- `getAnnotation()`/`pData()` (extracts a standard data frame containing the annotation)
- `getMetadata()`/`varMetadata()` (extracts a data frame of variable descriptions)
- `getVariableNames()`/`varLabels()` (extracts a vector of just the variable names)

You can also view a variable in one of these objects by using `$`, as you would for an ordinary data frame. You can also add new variables, but you need to be careful to update the metadata as well. Since we will be adding new variables in the process of cleaning our data, the following chunk of code shows you how to do this. Then you will practice looking at the annotation, and your task is to answer some questions about this data set.

First, load the annotation data (`annotation_data.RData`) that was provided with this project. **You need to supply the path to this file.**

```{r load_annotation, eval=TRUE}

# Load the probe (SNP) and sample annotation data frames, which have been saved in an .RData file
path_to_annotation_data <- "/ix1/hugen2072-2025s/p1/annotation_data.RData"
load(path_to_annotation_data, verbose = TRUE)

# List the loaded objects
ls()
typeof(scanAnnot); typeof(snpAnnot)

```

Here's a toy example showing you how to work with annotation data frames. **Read this carefully, since you'll need to do this throughout the assignment.**

```{r annotation_data_frame_example, eval=TRUE}

### How to add a new variable to an annotation data frame ###

#   Make a new copy of the SNP annotation (just for this example)
snp_annodf_temp <- snpAnnot

#   Notice that printing the object isn't that helpful
snp_annodf_temp

## Step1
#   So extract the annotation itself (this is the 'normal' data frame embedded
#   in the SNP annotation data frame object). Now you can see the SNP-level
#   annotation
snp_data_temp <- getAnnotation(snp_annodf_temp)
head(snp_data_temp)

#   Add a new column to the annotation (an integer dummy variable)
snp_data_temp$dummy <- 1:nrow(snp_data_temp)

#   Now update the scan annotation object itself by replacing the what was in
#   the annotation "slot"
pData(snp_annodf_temp) <- snp_data_temp

#   You can see that the new column has been added
head(pData(snp_annodf_temp))


## Step2
#   Extract the metadata so that it can be updated
meta_temp <- getMetadata(snp_annodf_temp)
meta_temp

#   Add a new row titled "dummy" with an appropriate labelDescription (note that
#   the metadata is a data frame with named rows but one column)
meta_temp["dummy", "labelDescription"] <- "This is a dummy variable"

#   Finally, replace the old metadata with the new metadata
varMetadata(snp_annodf_temp) <- meta_temp


## Step3
#   Look at the data and metadata to make sure there the dummy variable has been
#   added
head(getAnnotation(snp_annodf_temp))
getMetadata(snp_annodf_temp)

#   Remove the temp files
rm(snp_annodf_temp, snp_data_temp, meta_temp)

```

**Conclusion**: update Annotation data -> update metadata

<hr>

Now we use `GdsGenotypeReader()` and `GdsIntensityReader()` to open the gds files that contain the array data. Then use `GenotypeData()` and `IntensityData()` to combine them with the SNP and sample annotation, since the functions we'll be using later require this information to be "packaged" together.

```{r load_array_data, eval=TRUE}
##1
#   Get the genotypes from a gds file
genofile <- system.file("extdata", "illumina_geno.gds", package = "GWASdata")
gds_geno <- GdsGenotypeReader(genofile)

#   Combine the genotypes with the sample and SNP annotation
genoData <- GenotypeData(gds_geno, snpAnnot = snpAnnot, scanAnnot = scanAnnot)
genoData


##2.1
#   Get the intensity data from a gds file
qxyfile <- system.file("extdata", "illumina_qxy.gds", package = "GWASdata")
gds_qxy <- GdsIntensityReader(qxyfile)

#   Combine with the sample and SNP annotation
qxyData <- IntensityData(gds_qxy, snpAnnot = snpAnnot, scanAnnot = scanAnnot)
qxyData

##2.2
#   Get the BAF/LRR data from a gds file
blfile <- system.file("extdata", "illumina_bl.gds", package = "GWASdata")
gds <- GdsIntensityReader(blfile)

#   Combine the BAF/LRR data with the sample and SNP annotation
blData <- IntensityData(gds, snpAnnot = snpAnnot, scanAnnot = scanAnnot)
blData

```

## Part 0 - Report

Using the functions introduced above to help you, briefly describe the dataset in a few sentences. Answer the following questions. Whatever code you use to help answer these questions can go in the empty chunk that follows.

* What is the difference between the scanID and subjectID variables? Why are they both provided?

* How many families are represented, and how many samples are from each family?

* How many samples are of each sex?

* How many samples are from each population group?

* How many samples are there? How many people did they come from?

* How many SNPs are there?

* Which chromosomes are represented? How many SNPs are on each?

<br>

**Part 0 Report**

In Part 0, we imported annotation data (`scanAnnot` & `snpAnnot`) and array data (genotyping data and intensity data, including qxy and bl) into the work space and then combined/packed them for downstream QC processes. The combined data include `genoData` (genotype data), `qxyData`, and `blData` (intensity data).

Q1: In detail, we found both `scanID` and `subjectID` in Scan Annotation data. They are both provided because `subjectID` refers to a specific participant, while `scanID` refers to a specific genotyping run. It is highly possible for a subject/participant to be sequenced/genotyped for more than once, for intended duplication purpose or for different tissue types, etc. Thus, both ID should be provided. 

Q2: 13 families were included in the study, and the number of samples in each family is summarized in table `samples_per_fam`

Q3: Among all samples/scans, 35 were from females, and 42 were from males.

Q4: 49 samples were from CEU, and 28 samples were from YRI

Q5: There were 77 samples, which came from 43 people.

Q6: There were 3300 SNPs with unique `snpID` (and `rsID`)

Q7: Chromosomes included in this study:
chromosome  number_of_SNPs
21	        1000			
22	        1000			
23(X)	      1000			
24(Y)       100			
25(XY)      100			
26(MT)      100	

```{r part0_report}
# Your code goes here
## Q1
pData(scanAnnot) %>% head()

## Q2
pData(scanAnnot) %>% select(family) %>% unique() %>% nrow()

samples_per_fam = 
  pData(scanAnnot) %>% 
  select(family, scanID) %>% 
  group_by(family) %>%
  count()
samples_per_fam

## Q3
pData(scanAnnot) %>% select(sex, scanID) %>%
  group_by(sex) %>%
  count()

## Q4
pData(scanAnnot) %>% select(race, scanID) %>%
  group_by(race) %>%
  count()

## Q5
pData(scanAnnot) %>% select(scanID) %>% unique() %>% nrow()
pData(scanAnnot) %>% select(subjectID) %>% unique() %>% nrow()

## Q6
pData(snpAnnot) %>% select(snpID) %>% unique() %>% nrow()
# pData(snpAnnot) %>% select(rsID) %>% unique() %>% nrow()

## Q7
pData(snpAnnot) %>% select(chromosome, snpID) %>% group_by(chromosome) %>% count()

```

<hr>

<br>

## Part 1 - Missing Call Rate for Samples and SNPs

The basic `{GWASTools}` functions for missingness are `missingGenotypeBySnpSex()` and `missingGenotypeByScanChrom()`. 

- `missingGenotypeBySnpSex()` returns a list of three objects: `missing.counts`, `scans.per.sex`, and `missing.fraction`. Read the help file for this function so that you understand what information is contained in these and how the Y chromosome is treated by sex (where sex is as defined in the scan annotation). 

- Similarly, `missingGenotypeByScanChrom()` returns `missing.counts`, `snps.per.chr`, and `missing.fraction`.


There are four basic missingness calculations to be done, in the following order.

1. Calculate `missing.n1` = fraction of **genotype calls** missing over all samples (except that annotated ‘female’ individuals are excluded for Y chr SNPs)

2. Calculate `missing.e1` = missing call rate **per sample** over all SNPs (excluding SNPs with all calls missing)

3. Calculate `missing.n2` = call rate **per SNP** over all samples whose `missing.e1` is less than `0.05`

4. Calculate `missing.e2` = missing call rate **per sample** over all SNPs with `missing.n2` values less than `0.05`

We'll want to store the missingness information in the annotation data frames as we proceed.

<br>

### Round 1

Here's how to calculate and store `missing.n1`.

```{r missing.n1, eval=TRUE}

### Using missingGenotypeBySnpSex to calculate missing.n1 ###

#   First look at some of the results
miss <- missingGenotypeBySnpSex(genoData)
names(miss)

head(miss$missing.counts) #all snps
miss$scans.per.sex
head(miss$missing.fraction)

#   We should make sure the snpIDs are in the same order in the annotation and
#   in the missingness report (miss$missing.fraction); they are

allequal(snpAnnot$snpID, as.numeric(names(miss$missing.fraction)))

#   Then go ahead and add missing.n1 as a new annotation data column, and update
#   the metadata
pData(snpAnnot)$missing.n1 <- miss$missing.fraction
varMetadata(snpAnnot)["missing.n1", "labelDescription"] <- 
  paste("fraction of genotype calls missing over all samples, except that females are excluded for Y chr SNPs")

## checks
getVariableNames(snpAnnot)
getMetadata(snpAnnot) %>% tail(n=2)

### Explore the missing call rate per probes a little more

#   Look at the distribution of missingness for all the probes
hist(
  pData(snpAnnot)$missing.n1,
  ylim = c(0, 100),
  xlab = "SNP missing call rate",
  main = "Missing Call Rate for All Probes"
)

#   How many SNPs are missing a genotype for every single sample? Apparently
#   there are 151 such SNPs

sum( pData(snpAnnot)$missing.n1 == 1 )

#   Store a vector containing the snpIDs of all the SNPs with missingness of
#   100%

snpexcl <- pData(snpAnnot)$snpID[snpAnnot$missing.n1 == 1]

```

<br>

Now we calculate and store `missing.e1`. The steps are described in the comments below.

```{r missing.e1, eval=TRUE}

### Using missingGenotypeByScanChrom to calculate missing.e1

#   Apply missingGenotypeByScanChrom to genoData, using the snp.exclude option
#   to exclude these SNPs (why does it make sense to exclude those SNPs?)

miss <- missingGenotypeByScanChrom(genoData, snp.exclude = snpexcl)

#   Make sure the order of the scanIDs in the output from
#   missingGenotypeByScanChrom is the same as the order of the scanIDs in the
#   annotation data frame

allequal(names(miss$missing.fraction) %>% as.numeric(), 
         pData(scanAnnot)$scanID)

#   Make a new column in scanAnnot called missing.e1 and store the per-scan
#   missingness from missingGenotypeByScanChrom there

pData(scanAnnot)$missing.e1 <- miss$missing.fraction

#   Update the scanAnnot metadata - using setter functions

varMetadata(scanAnnot)["missing.e1", "labelDescription"] =
  "fraction of genotype calls missing over all snps with missing.n1<1 except that Y chr SNPs are excluded for females"


## check
getVariableNames(scanAnnot)
getMetadata(scanAnnot) %>% tail(n=2)


# We can check summary statistics and make a histogram, too

summary(scanAnnot$missing.e1)
hist(
  scanAnnot$missing.e1,
  xlab = "Fraction of sample-missingness across all probes",
  main = "Histogram of Sample Missing Call Rate for All Samples"
)

```

<br>

### Round 2

Now, supply your code to calculate and store `missing.n2`. The steps you should follow are described in the comments below. 

```{r missing.n2, eval=TRUE}

### Now that missing.n1 and missing.e1 have been calculated, we can calculate missing.n2

#   Recall that missing.n2 = call rate per SNP over all samples whose missing.e1
#   is less than 0.05.

#       Since no sample had missing.e1 > 0.05 in this dataset, we don't actually
#       have to exclude any samples.

#   However, let's proceed as if there were some samples with missing.e1 > 0.05.

# Make a vector of the scanIDs for the samples with high missing.e1.
scan_exclude = 
  pData(scanAnnot) %>% 
  filter(missing.e1 > 0.05) %>%
  select(scanID) %>% 
  pull()

length(scan_exclude)

#   Apply the missingGenotypeBySnpSex function again, using the scan.exclude
#   argument.
miss = missingGenotypeBySnpSex(genoData = genoData, scan.exclude = scan_exclude)

#   Store missing.n2 as a new column in snpAnnot.
## same sequence
all.equal( names(miss$missing.fraction) %>% as.numeric(), 
           pData(snpAnnot)$snpID ) #TRUE

## assign to annot data
pData(snpAnnot)$missing.n2 = miss$missing.fraction

#   Update the metadata for snpAnnot!
varMetadata(snpAnnot)["missing.n2", "labelDescription"] = "SNP-level missingness, in samples with sample-missing rate (missing.e1) <= 5%"

#   The labelDescription for missing.n2 should be "fraction of genotype calls
#   missing over all samples with missing.e1 < 0.05 except that females are
#   excluded for Y chr SNPs"

## checks
getVariableNames(snpAnnot)
getMetadata(snpAnnot) %>% tail(n=2)

```

<br>

Write code to calculate and store `missing.e2`. The steps you should follow are described in the comments below.

```{r missing.e2, eval=TRUE}

### Last, we calculate missing.e2

#   Recall missing.e2 = missing call rate per sample over all SNPs with
#   missing.n2 values less than 0.05

#   Store a vector with the snpIDs of all probes with missing.n2 >= 0.05
snp_exclude_r2 = 
  pData(snpAnnot) %>%
  filter(missing.n2 >= 0.05) %>%
  pull(snpID) #snps with missing>5%

length(snp_exclude_r2)

#   Apply missingGenotypeByScanChrom to genoData again, using the snp.exclude
#   argument to exclude the SNPs with missing.n2 >= 0.05 (why do we do this?)

miss = missingGenotypeByScanChrom(genoData, snp.exclude = snp_exclude_r2)
  ## We do this to ensure that the bad snps are removed prior to removing bad samples, which has a larger effect on downstream analysis.


#   Make a new column in scanAnnot called missing.e2 to store the missingness
#   per-scan you just calculated
all.equal( names(miss$missing.fraction) %>% as.numeric(), 
           pData(scanAnnot)$scanID ) #same sequence

pData(scanAnnot)$missing.e2 = miss$missing.fraction

#   Update the scanAnnot metadata
varMetadata(scanAnnot)["missing.e2","labelDescription"] = "Sample-missingness across all snps with missing.n2 < 0.05, except that Y chr SNPs are excluded for females"

## checks
getVariableNames(scanAnnot)
getMetadata(scanAnnot) %>% tail(n=2)

```

## Part 1 - Report

Summarize the filtering you've accomplished at this step. Include the following in your answer the following questions.

* (DONE) (First, make sure you've filled in the calculations of `missing.n2` and `missing.e2` in the chunks above.)

* (DONE) How many SNPs and samples were genotyped?

* (DONE) Briefly describe the purpose of doing this step in two rounds (why do we calculate both missingnesses twice?)

* (DONE) What were the median, mean, and maximum per-sample missingness (use `missing.e2`)? Make a histogram of `missing.e2`.

* (DONE) How many samples had no missing calls at all (use `missing.e2`)?

* (DONE) Are the samples generally "good"? How many should be excluded from downstream analyses?

* (DONE) How many SNPs had a 100% missing call rate (use `missing.n1`)?

* (DONE) What are the median, mean, and maximum per-SNP missingnesses (use `missing.n2`)? How many SNPs are missing no calls at all?

* (DONE) Make a histogram of `missing.n2`.

* (DONE) How many SNPs are filtered out and how many remain if you retain only SNPs with `missing.n2` < 0.05?

* (DONE) Make a histogram showing `missing.n2` for the remaining SNPs.

<br>

**Report**: We have finished 2 rounds of missingness checks. In this study, 3300 SNPs and 77 samples were genotyped. Our rationale of performing 2 rounds of missingness check is that for the first round, we try to capture the problematic SNPs (totally missed) and, after excluding them, the problematic samples with high missing rate. Based on the primarily filtered data, we can further refine the missingness calculation in the second round, which will be used in the downstream pipeline.

The median, mean, and max sample-missingness were calculated to be 0.00065, 0.0018, and 0.024, respectively. The histogram for `missing.e2` is shown below, which is named as "Histogram of sample-missingness (missing.e2) across high-quality SNPs". There were 6 samples with no missing calls at all (i.e., sample-level missingness = 0). *Sample-level missing rate < 5%* were defined as generally "good", and all samples met this criteria and no samples should be excluded in the downstream analyses. 

There were 151 SNPs with 100% missing call rate. For the SNP-level missingness, we have median=0, mean=0.049, max=1. 2755 SNPs were missing no calls at all (i.e., snp-level missingness = 0). We created a histogram for `missing.n2` below, which is named "Histogram of SNP-level missingness (missing.n2) across high-quality samples". Using a filtering threshold of *SNP-level missingness < 0.05*, we kept 3095 SNPs while removing 205 SNPs. The histogram for the remaining SNPs were generated, named "Histogram of SNP-level missingness for high-quallity SNPs across high-quality samples".


```{r part1_report}
# check for missing.n2 & missing.e2
# pData(snpAnnot)$missing.n2
# pData(scanAnnot)$missing.e2

pData(snpAnnot)$snpID %>% unique() %>% length()
pData(scanAnnot)$scanID %>% unique() %>% length()

# What were the median, mean, and maximum per-sample missingness (use `missing.e2`)? Make a histogram of `missing.e2`.
summary( pData(scanAnnot)$missing.e2 ) 
hist(pData(scanAnnot)$missing.e2, 
     ylim = c(0, 100), 
     xlab = "Sample missingness",
     main = "Histogram of sample-missingness (missing.e2)\n across high-quality SNPs")

# How many samples had no missing calls at all (use `missing.e2`)?
sum( pData(scanAnnot)$missing.e2 == 0 )

# Are the samples generally "good"? How many should be excluded from downstream analyses?
sum( pData(scanAnnot)$missing.e2 >= 0.05 )

# How many SNPs had a 100% missing call rate (use `missing.n1`)?
sum( pData(snpAnnot)$missing.n1 == 1 )

# What are the median, mean, and maximum per-SNP missingnesses (use `missing.n2`)? How many SNPs are missing no calls at all?
summary( pData(snpAnnot)$missing.n2 ) #median=0, mean=0.049, max=1
sum( pData(snpAnnot)$missing.n2 == 0 )

# Make a histogram of `missing.n2`
hist(pData(snpAnnot)$missing.n2,
     xlab = "SNP-level missingness",
     main = "Histogram of SNP-level missingness (missing.n2)\n across high-quality samples")

# How many SNPs are filtered out and how many remain if you retain only SNPs with `missing.n2` < 0.05?
## kept
sum( pData(snpAnnot)$missing.n2 < 0.05 ) #3095 kept
sum( pData(snpAnnot)$missing.n2 >= 0.05 ) #205 lost
# length(snp_exclude_r2)

# Make a histogram showing `missing.n2` for the remaining SNPs.
pData(snpAnnot) %>%
  filter( !(snpID %in% snp_exclude_r2) ) %>%
  # summarize(tmp=max(missing.n2)) #check
  pull(missing.n2) %>%
  hist(., xlab="SNP-level missingness",
       main="Histogram of SNP-level missingness\n for high-quallity SNPs across high-quality samples")

```

<hr>

<br>

## Part 2 - Inferred vs. Reported Sex and Relatedness

To investigate discrepancies in inferred vs. reported sex (and, possibly, sex chromosome aneuploidies), we will produce four plots that summarize each scan's heterozygosity and intensity for the sex chromosomes:

1. Mean Y chromosome vs. mean X chromosome intensity

2. Mean X chromosome heterozygosity vs. mean X chromosome intensity

3. Mean Y chromosome intensity vs. mean X chromosome heterozygosity

4. Mean X chromosome heterozygosity vs. autosomal heterozygosity

Recall that the intensity data was read in earlier as `qxyData`. Before making our graphs, we'll need to calculate the variables we want to plot. The basic function we'll use to do this is `meanIntensityByScanChrom()`, which will return a list of matrices. Read the help file of this function so that you understand the output. Heterozygosity is calculated with `hetByScanChrom()`.

```{r sex_check, eval=TRUE}

#   Look at the structure of the IntensityData object
qxyData
qxyData@data

#   For each probe, there's a numerical ID, chromosome, position, rsID

#   The intensities (stored as variables called X and Y) are each essentially
#   3300×77 matrices (an X and Y intensity for each scan for every single probe)

#   (Don't confuse the *intensity variable names* _X_ and _Y_ in the IntensityData
#   object with the X and Y *chromosomes*!)

#   Apply the meanIntensityByScanChrom function to get the intensity averages.
#   (We're not going to use all of the output from this function)
inten.by.chrom <- meanIntensityByScanChrom(qxyData)

#   Extract the matrix containing the mean intensity for each chromosome for each scan
names(inten.by.chrom)
mninten <- inten.by.chrom$mean.intensity

#   The matrix is 77x6 since there are 77 samples and 6 chromosomes represented in the dataset
dim(mninten)
head(mninten)


#   We also need to calculate the X chromosome and autosome heterozygosities for
#   each sample before we can plot them

#   Apply hetByScanChrom to genoData (we need actual genotypes to calculate the
#   heterozygosity)
het.results <- hetByScanChrom(genoData)
    #   The function calculates eachs scan's heterozygosity for each chromosome, and
    #   also for the autosomes as a whole

#   We want to add X and autosome heterozygosity as new columns in scanAnnot
#   Make sure the scanIDs are in the right order first! (they are)
allequal(scanAnnot$scanID, rownames(het.results))

#   Extract the autosomal and X heterozygosity and store them in scanAnnot
scanAnnot$het.A <- het.results[, "A"]
scanAnnot$het.X <- het.results[, "X"]

#   Add appropriate metadata
varMetadata(scanAnnot)["het.A", "labelDescription"] <- "fraction of heterozygotes for autosomal SNPs"
varMetadata(scanAnnot)["het.X", "labelDescription"] <- "fraction of heterozygotes for X chromosome SNPs"


#   Before making the plots, there are a few things to do: We want to color the plotted points by annotated sex

#   But first, make sure that the order of the scanIDs (and hence the annotated sexes) matches the order of the scanIDs in the data we'll plot
allequal(scanAnnot$scanID, rownames(mninten))

#   Now we assign each sex a color (male=blue, female=red) - Store the colors in a vector
xcol <- rep(NA, nrow(scanAnnot))
xcol[scanAnnot$sex == "M"] <- "blue"
xcol[scanAnnot$sex == "F"] <- "red"

#   We also want to count the number of SNPs getting included on the X and Y chromosomes (which are coded as 23 and 25 in snpAnnot)
nx <- sum(snpAnnot$chromosome == 23)
ny <- sum(snpAnnot$chromosome == 25)


#### Now make the plots - Setup parameters for each plot

# All intensities
x1 <- mninten[, "X"]
y1 <- mninten[, "Y"]
main1 <- "Mean X vs \nMean Y Chromosome Intensity"

# Het on X vs X intensity
x2 <- mninten[, "X"]
y2 <- scanAnnot$het.X
main2 <- "Mean X Chromosome Intensity vs Mean X Chromosome Heterozygosity"

# Het on X vs Y intensity
y3 <- mninten[, "Y"]
x3 <- scanAnnot$het.X
main3 <- "MeanX Chromosome Heterozygosity vs Mean Y Chromosome Intensity"

# X vs A het
x4 <- scanAnnot$het.A[scanAnnot$sex == "F"]
y4 <- scanAnnot$het.X[scanAnnot$sex == "F"]
main4 <- "Mean Autosomal Heterozygosity vs Mean X Chromosome Heterozygosity"


# Make labels for axes/legends
cols <- c("blue", "red")
mf <- c("male", "female")
xintenlab <- paste("X intensity (n=", nx, ")", sep = "")
yintenlab <- paste("Y intensity (n=", ny, ")", sep = "")


#### Make the 4 plots and add a legend
par(mfrow = c(2, 2))
plot(
  x1,
  y1,
  xlab = xintenlab,
  ylab = yintenlab,
  main = main1,
  col = xcol,
  cex.main = 0.8
);legend("bottomleft",
       mf,
       col = cols,
       pch = c(1, 1),
       cex = 0.5)
plot(
  x2,
  y2,
  col = xcol,
  xlab = xintenlab,
  ylab = "X heterozygosity",
  main = main2,
  cex.main = 0.8
)
plot(
  x3,
  y3,
  col = xcol,
  ylab = yintenlab,
  xlab = "X heterozygosity",
  main = main3,
  cex.main = 0.8
)
plot(
  x4,
  y4,
  col = "red",
  xlab = "Autosomal heterozygosity",
  ylab = "X heterozygosity",
  main = main4,
  cex.main = 0.8
)
par(mfrow = c(1, 1))

```


## Part 2 - Report (Sex Check)

* (DONE) Describe any anomalies you see in the 4 plots above.

* (DONE) Figure out the scanIDs of any scans that look like outliers or possible instances of mislabeled sex. Include these IDs in your report and describe why they appear to be mislabeled.

* (DONE) For the purposes of this assignment we will ASSUME that some scans have just had their sexes mislabeled. (In "real life", we would not just assume this, and we might instead flag these scans for removal from some downstream steps/analyses.)

* (DONE) Manually change the sex annotation for these scans in scanAnnot to "remove" the discrepancy. (Manually changing your data is risky and not generally recommended. Be careful!). 
    * (DONE) In your report, state exactly what changes you made. Add a new column to scanAnnot indicating this, too (add appropriate metadata). Does anything in the pedigree structure make you more confident that this "fix" is correct? (Are these samples labeled as mothers or fathers?)

* (DONE) Remake the 4 plots above, using the "corrected" data (nothing should appear mislabeled now for this dataset).

* (DONE) How might you follow up on this "in real life"?

<br>

**Report**:  (side-note:`scanAnnot`, `mninten`)

From the chrY vs chrX intensity plot, we found 2 scans (scanID=325,326) whose reported sex was female, but they showed strong Y intensity but lower X intensity, suggesting a mismatch here and that their inferred sex is male. 

From the chrX het vs chrX intensity plot, we found that in the males cluster, there were reported females (scanID=325,326) showing low X intensity and X het=0, which should be inferred as a male. The scanID matches with plot 1.

From the chrY intensity vs chrX het plot, we found that in the males cluster, there were reported females (scanID=325,326) showing het.x=0 and with high Y intensity, which should be inferred as males. (same information as previous two plots)

From the chrX het vs autosomal het plot, we found outliers & reported females who has 0 chrX heterozygosity (scanID=325,326), which should be inferred as males. Since autosomal heterozygosity should be a spectrum, I am not very worried about the outliers on the x-axis, but to *keep a note the two reported females scanID=338,339*.

- [Conclusion] wrong reported sex: scanID = 325, 326 (should be males)

We further manually corrected the mislabeled sex for scanID=325,326 in the scanAnnot by the code: `pData(scanAnnot)[which( pData(scanAnnot)$scanID %in% c(325, 326) ),]$sex = rep(unique(pData(scanAnnot)$sex)[2], 2)`, which is to change the sex from "F" to "M". We further added a new column `changed_to_QC_inferredSex` (only 325, 326 is TRUE) to scanAnnot, indicating this change. From the pedigree structure of family 1408, the subjectID (200094287) associated with the scanIDs is labeled as **father** of the family, further indicating that our fix should be correct.

After remaking the 4 plots using data with updated sex, there are no more mismatches observed. 

In real life, we would further check the BAF and LRR plots to confirm if the copy number or intensity of the scans' sex chromosomes are expected. Also, we can check on the plate, batch, and other technical information logged in the data to see if there are other possibilities contributing to this mismatch. 


```{r part2_report_sex_check}
# Plot 1
ls_plt1_reportedSex = 
    pData(scanAnnot) %>%
    select(scanID, sex) %>%
    filter(sex=="F") %>% #find those reported as females
    pull(scanID)

ls_plt1_reportedSex[
    #find those inferred as males -> find the overlap
    ls_plt1_reportedSex %in% rownames(mninten)[mninten[,"Y"] > 0.5] 
]

# Plot 2
pData(scanAnnot) %>%
    select(scanID, sex, het.X) %>%
    filter(sex=="F" & het.X < 0.15) %>% #find those reported as females && low het.x
    pull(scanID)

c(325,326) %in% rownames(mninten)[mninten[,"X"]<0.9] #confirmed

# Plot 3 - same info as before
c(325,326) %in% rownames(mninten)[mninten[,"Y"] > 0.5]

# Plot 4 - same info as before
pData(scanAnnot) %>%
    select(scanID, sex, het.A) %>%
    filter(sex=="F" & het.A < 0.282) %>% #find those reported as females && low het.A
    pull(scanID)
```


Manually change the mislabeled sex

```{r}
# 1. Update sex
pData(scanAnnot)[which( pData(scanAnnot)$scanID %in% c(325, 326) ),]$sex =
    rep(unique(pData(scanAnnot)$sex)[2], 2)
## check
pData(scanAnnot)[which( pData(scanAnnot)$scanID %in% c(325, 326) ),]$sex


# 2. Add a new column to indicate the change
pData(scanAnnot) = 
    pData(scanAnnot) %>%
    mutate(changed_to_QC_inferredSex = 
               if_else(condition=scanID %in% c(325, 326), 
                       true=TRUE, false=FALSE) 
           ) #%>% filter(changed_to_QC_inferredSex)
## checks
pData(scanAnnot) %>% filter(changed_to_QC_inferredSex)


# 3. Modify the metaData
varMetadata(scanAnnot)["changed_to_QC_inferredSex","labelDescription"] = 
    "Logical vector; TRUE if we manually changed the scanID's `sex` to the opposite, based on the results of QC step 'Reported Sex vs Inferred Sex' "
## checks
varMetadata(scanAnnot) %>% tail(n=2)
getVariableNames(scanAnnot)


# 4. check pedigree
pData(scanAnnot) %>% filter(family==1408)

```

Remake the 4 plots above, using the "corrected" data 

```{r}
allequal(scanAnnot$scanID, rownames(mninten))

# Reassign sex
xcol <- rep(NA, nrow(scanAnnot))
xcol[scanAnnot$sex == "M"] <- "blue"
xcol[scanAnnot$sex == "F"] <- "red"

#   We also want to count the number of SNPs getting included on the X and Y chromosomes (which are coded as 23 and 25 in snpAnnot)
nx <- sum(snpAnnot$chromosome == 23)
ny <- sum(snpAnnot$chromosome == 25)


#### Now make the plots - Setup parameters for each plot

# All intensities
x1 <- mninten[, "X"]
y1 <- mninten[, "Y"]
main1 <- "[updated] Mean X vs \nMean Y Chromosome Intensity"

# Het on X vs X intensity
x2 <- mninten[, "X"]
y2 <- scanAnnot$het.X
main2 <- "[updated] Mean X Chromosome Intensity vs Mean X Chromosome Heterozygosity"

# Het on X vs Y intensity
y3 <- mninten[, "Y"]
x3 <- scanAnnot$het.X
main3 <- "[updated] MeanX Chromosome Heterozygosity vs Mean Y Chromosome Intensity"

# X vs A het
x4 <- scanAnnot$het.A[scanAnnot$sex == "F"]
y4 <- scanAnnot$het.X[scanAnnot$sex == "F"]
main4 <- "[updated] Mean Autosomal Heterozygosity vs Mean X Chromosome Heterozygosity"


# Make labels for axes/legends
cols <- c("blue", "red")
mf <- c("male", "female")
xintenlab <- paste("X intensity (n=", nx, ")", sep = "")
yintenlab <- paste("Y intensity (n=", ny, ")", sep = "")


#### Make the 4 plots and add a legend
par(mfrow = c(2, 2))
plot(
  x1,
  y1,
  xlab = xintenlab,
  ylab = yintenlab,
  main = main1,
  col = xcol,
  cex.main = 0.8
);legend("bottomleft",
       mf,
       col = cols,
       pch = c(1, 1),
       cex = 0.5)
plot(
  x2,
  y2,
  col = xcol,
  xlab = xintenlab,
  ylab = "X heterozygosity",
  main = main2,
  cex.main = 0.8
)
plot(
  x3,
  y3,
  col = xcol,
  ylab = yintenlab,
  xlab = "X heterozygosity",
  main = main3,
  cex.main = 0.8
)
plot(
  x4,
  y4,
  col = "red",
  xlab = "Autosomal heterozygosity",
  ylab = "X heterozygosity",
  main = main4,
  cex.main = 0.8
)
par(mfrow = c(1, 1))
```

<hr>

<br>

## Part 3 - Allelic Imbalance; BAF and LRR Plots

Here we want to look at B allele frequency (BAF) and log R ratio (LRR) plots using `chromIntensityPlot()`. Ordinarily, since there are 22 to 23 chromosomes to check for every sample, you would use a script that automatically detects aberrations in BAF and LRR. For this project, we'll pretend that our pipeline has already flagged one sample for a possible chromosomal anomaly.

```{r baflrr, eval=TRUE}

# Recall that earlier we read the intensity data into an IntensityData object
# called blData.

# The snp/scan annotation were attached as well.

# Suppose that chromosome 22 was flagged for scanID 286.

# Making the BAF/LRR plot is easy

#dev.off()
chromIntensityPlot(blData, scan.ids = 286, chrom.ids = 22)
par(mfrow = c(1, 1))

```

## Part 3 - Report

* Describe the purpose of this step.

* Show an example of a "normal" pair of BAF/LRR plots (choose any chromosome/scanID you want).

* How would you detect trisomy 21 (i.e., describe briefly in words what the BAF/LRR plot would look like).

* Suppose an individual had a balanced reciprocal translocation involving chromosome 2 and 4. What would you expect that individual's BAF/LRR plots would look like for those chromosomes?

* Suppose chromosome 22 in sample 286 was the only chromosome flagged for a possible anomaly in this dataset. Looking at the plot, do you notice anything strange? Can you suggest what kind of anomaly might be showing up? (There isn't any single answer that's certainly right! The purpose of this question is to check that you understand what the BAF and LRR plots "mean").

* Finally, suppose your pipeline has given you a list of the exact chromosomal segments showing possibly aneuploidies/anomalies in each sample. What would you recommend doing with the genotypes in those segments? (You're not being asked to implement this here - just say what you recommend doing.)

```{r part3_report}

#  <Your code goes here>

```
